pkg/util/cloudfoundry/bbscache.go

```diff
@@ -0,0 +1,274 @@
+func (bc *BBSCache) GetAllLRPs() (map[string][]ActualLRP, map[string]DesiredLRP) {
+	bc.RLock()
+	defer bc.RUnlock()
+	return bc.actualLRPs, bc.desiredLRPs
+}
+
+func (bc *BBSCache) start() {
+	bc.readData()
+	dataRefreshTicker := time.NewTicker(bc.pollInterval)
+	for {
+		select {
+		case <-dataRefreshTicker.C:
+			bc.readData()
+		}
+	}
+}
+
+func (bc *BBSCache) readData() {
+	log.Debug("Reading data from BBS API")
+	bc.Lock()
+	bc.pollAttempts++
+	bc.Unlock()
+	var wg sync.WaitGroup
+	var actualLRPs map[string][]ActualLRP
+	var actualLRPByInstanceGUID map[string]*ActualLRP
+	var desiredLRPs map[string]DesiredLRP
+	var errActual, errDesired error
+
+	wg.Add(2)
+
+	go func() {
+		actualLRPs, actualLRPByInstanceGUID, errActual = bc.readActualLRPs()
+		wg.Done()
+	}()
+	go func() {
+		desiredLRPs, errDesired = bc.readDesiredLRPs()
+		wg.Done()
+	}()
+	wg.Wait()
+	if errActual != nil {
+		log.Errorf("Failed reading Actual LRP data from BBS API: %s", errActual.Error())
+		return
+	}
+	if errDesired != nil {
+		log.Errorf("Failed reading Desired LRP data from BBS API: %s", errDesired.Error())
+		return
+	}
+
+	// put new values in cache
+	bc.Lock()
+	defer bc.Unlock()
+	log.Debug("Data from BBS API read successfully, refreshing the cache")
+	bc.actualLRPs = actualLRPs
+	bc.actualLRPByInstanceGUID = actualLRPByInstanceGUID
+	bc.desiredLRPs = desiredLRPs
+	bc.lastUpdated = time.Now()
+	bc.pollSuccesses++
+}
+
+func (bc *BBSCache) readActualLRPs() (map[string][]ActualLRP, map[string]*ActualLRP, error) {
+	actualLRPs := map[string][]ActualLRP{}
+	actualLRPByInstanceGUID := map[string]*ActualLRP{}
+	actualLRPsBBS, err := bc.bbsAPIClient.ActualLRPs(bc.bbsAPIClientLogger, models.ActualLRPFilter{})
+	if err != nil {
+		return actualLRPs, actualLRPByInstanceGUID, err
+	}
+	for _, lrp := range actualLRPsBBS {
+		alrp := ActualLRPFromBBSModel(lrp)
+		if lrpList, ok := actualLRPs[alrp.AppGUID]; ok {
+			actualLRPs[alrp.AppGUID] = append(lrpList, alrp)
+			actualLRPByInstanceGUID[alrp.InstanceGUID] = &actualLRPs[alrp.AppGUID][len(lrpList)]
+		} else {
+			actualLRPs[alrp.AppGUID] = []ActualLRP{alrp}
+			actualLRPByInstanceGUID[alrp.InstanceGUID] = &actualLRPs[alrp.AppGUID][0]
+		}
+	}
+	return actualLRPs, actualLRPByInstanceGUID, nil
+}
+
+func (bc *BBSCache) readDesiredLRPs() (map[string]DesiredLRP, error) {
+	desiredLRPsBBS, err := bc.bbsAPIClient.DesiredLRPs(bc.bbsAPIClientLogger, models.DesiredLRPFilter{})
+	if err != nil {
+		return map[string]DesiredLRP{}, err
+	}
+	desiredLRPs := make(map[string]DesiredLRP, len(desiredLRPsBBS))
+	for _, lrp := range desiredLRPsBBS {
+		desiredLRP := DesiredLRPFromBBSModel(lrp)
+		desiredLRPs[desiredLRP.AppGUID] = desiredLRP
+	}
+	return desiredLRPs, nil
+}
+
+// ExtractTags extract all the container tags for each app in the cache an returns a mapping of tags by instance GUID
+func (bc *BBSCache) ExtractTags() map[string][]string {
```

Comment:
would be super useful we can map tags by nodes (VMs), each node agent puts its node name in the query path and gets only a subset of tags of apps running on that same node